### ISSUE TITLE
`generate_trajectories()` is slow on a database with many users

### ISSUE THREAD

**Issue by asiripanich** [asiripanich/emdash#29]
Reported by @shankari. 

We should profile our app and see where the real issue is. The easiest way to fix this is to run `generate_trajectories()` in the background process, so that the UI doesn't freeze.

**Comment by shankari** [comment]
One disadvantage with running `generate_trajectories()` in the background instead of on demand, is that we incur the network bandwidth cost even if we only want to check the participants or plot the trips and not look at the map. Assuming people are going to run their server in the cloud, and since most cloud providers charge for bandwidth, this could add up quite quickly.

We should at least determine whether people would be willing to filter their trajectory queries, either by geography or by time before loading them.

**Comment by shankari** [comment]
Also FYI, I just tried to load the dashboard for the mini-pilot. ~25 users, but ~ 6 months of data. I think it has crashed the AWS server.

**Comment by shankari** [comment]
<img width="1065" alt="Screen Shot 2021-06-02 at 6 45 44 AM" src="https://user-images.githubusercontent.com/2423263/120491559-318bda80-c36e-11eb-8e72-cb1cea7138f0.png">

**Comment by asiripanich** [comment]
> 
> 
> One disadvantage with running `generate_trajectories()` in the background instead of on demand, is that we incur the network bandwidth cost even if we only want to check the participants or plot the trips and not look at the map. Assuming people are going to run their server in the cloud, and since most cloud providers charge for bandwidth, this could add up quite quickly.
> 
> We should at least determine whether people would be willing to filter their trajectory queries, either by geography or by time before loading them.

Can we check the size of a query before making it? Let's say if bigger than 100MB then warn the user that this query might crash their server?

So the steps would be:
1) Check the size of the query
2) Warn if exceeds `x`MB
3) Query

**Comment by shankari** [comment]
@asiripanich but what if it does exceed `x` MB? Is the dashboard unusable at that point?

**Comment by asiripanich** [comment]
We put a warning prompt and let them decide? 

Also, are you a MongoDB expert by any chance? :)

**Comment by shankari** [comment]
but a decision which says "your computer may crash" is not really a decision. What if I (like a sane person) don't want my computer to crash? Then I have only one choice which is to not use the dashboard.

**Comment by asiripanich** [comment]
Are you proposing that we should not allow a query to happen if it exceeds a limit? I'm ok either way, but some people may have a better spec machine than the others. :)

**Comment by shankari** [comment]
no, I'm proposing that we allow users to filter by time range and users (and potentially geo range).
So then if the query is too large, they can shorten the query parameters and still use the dashboard.

here's the discussion that @allenmichael099 had today:
- @allenmichael099: I profiled the code. It looks like the longest runtime part of generate_trajectories is when tidied_locations and tidied_trips are joined with foverlaps
- @shankari: cool. but there are also just a lot of trajectory points, right? can you look at an estimate per month? I bet is is an order of magnitude larger than any of the other objects. so the trip memory usage is O(n) but the point memory usage is O(n^3) or something. and so, over a few months, it is going to overwhelm RAM on a single computer. (which is likely what caused AWS to crash). and even otherwise, even an O(n) growth will eventually exhaust finite ram depending on the constant factor. depends on the ram and the time frame, but we are thinking about 2 years of data here
- @allenmichael099: How do you think we should cut back on the number of points? Look at specific months or a subset of users?
- @shankari: yup! we can load only participants on startup, filling in trip information using counts from the database instead of loading all trips and doing a merge. the participant table already has the range of trips for each user. then for loading both trips and trip trajectories, use a filter based on time and set of users

@asiripanich as you suggested, we can do a `count()` for the filter results before actually retrieving the results.

**Comment by asiripanich** [comment]
Certainly, I wasn't suggesting that we shouldn't allow filter by time range and users. I just thought that we should also check for the size of the query before making it. MongoDB `count()` only counts the number of entries right but not its file size?

**Comment by shankari** [comment]
right. I don't see an obvious way to get the size (MB) of a query at runtime. But we know the rough size of the objects, so we can estimate the size the query result from the count. mongodb does have methods to return average object size in collection, etc. so I think that is what they expect.

**Comment by shankari** [comment]
Even if we switch to filtering, we can filter on the R server or on the database.
Obviously, filtering on the database is more performant because of the built-in indices.
https://github.com/grahamrp/dtdatasources#about-custom-dt-data-sources


Some links for database level filtering:
https://github.com/rstudio/DT/issues/194
https://github.com/grahamrp/dtdatasources

**Comment by shankari** [comment]
Since the bottleneck is clearly memory, we should query for subsets of the data from the database. (https://github.com/e-mission/e-mission-docs/issues/645)
Three possible restrictions are:
(1) query for a time range (e.g. two weeks at a time, or a month at a time)
(2) query for a particular region (this may not be that useful for the CEO project because we have one container per city)
(3) downsample the timeseries entries (if we are viewing trajectories across an entire city, we don't need to show points every 30 secs or so, it might be enough to show points every few minutes). If we show every 5 mins for example, this will reduce the query size by 10x

Items (2) and (3) are related. If a user zooms in to a neighbourhood, for example, we could sample at 30 secs because there won't be as many trips to begin with.

**Comment by shankari** [comment]
We should start with (1) since it is low hanging fruit.
(3) is likely to be the next easy step because downsampling is supported automatically in mongo.
https://docs.mongodb.com/manual/reference/operator/aggregation/sample/

**Comment by allenmichael099** [comment]
Currently I'm working on querying the trips data using timestamps based on the user's selected date range. 

I have successfully queried by timestamp, but when I try to convert the resulting data frame to a data table with as.data.table, I get this error: 
```Error in `[[<-.data.frame`(`*tmp*`, i, value = list()) : 
  replacement has 0 rows, data has 22```

The error happens only when I reduce the date range from the full date range.

```
devtools::load_all('.../emdash/')
cons <- connect_stage_collections(url = getOption("emdash.mongo_url"))

# Query trips by timestamp
# dates should be a vector of two dates: start date and end date
query_cleaned_trips_by_timestamp <- function(cons,dates) {
  
  # Convert the dates to timestamps
  time_stamps <- as.numeric(as.POSIXct(dates))
  
  # Construct the query string.
  lower_stamp_string <- paste0('{\"$gte\": ',time_stamps[1], ',')
  upper_stamp_string <- paste0('\"$lte\": ',time_stamps[2], '}')
  
  qstring <- paste0('{\"metadata.key\": \"analysis/confirmed_trip\", ' ,
                    '\"data.end_ts\":' , lower_stamp_string, upper_stamp_string,
                    '}')
  
  # The query string should have this format
  # {"metadata.key": "analysis/confirmed_trip", "data.end_ts":{"$gte": 1437544800,"$lte": 1451286000}}
  
  cons$Stage_analysis_timeseries$find(qstring)

}

dates_all <- c("2015-07-22","2016-08-11")
dates_subset <- c("2015-07-22" , "2015-12-28") # These correspond to time_stamps <- c(1437544800, 1451286000)

# Works fine: 
trips <- query_cleaned_trips_by_timestamp(cons,dates_all) %>% as.data.table()

# Throws error:
trips_subset <- query_cleaned_trips_by_timestamp(cons,dates_subset) %>% as.data.table()
```

**Comment by allenmichael099** [comment]
Also I'm using test-data in this example. The metadata and data fields for both queries have the same number of columns, so I'm not sure where the difference leading to the error might be. What are your thoughts @asiripanich?

**Comment by asiripanich** [comment]
Now we know that `trips_subset` didn't work because they do not have user_input. 

```r
tidy_cleaned_trips_by_timestamp <- function(df) {
    dt <- df[, c("user_id", "metadata")] %>% as.data.table()
    data_df <- df[, c("data")]
    data_df_without_user_input <- data_df[, !"user_input" == names(data_df)]
    user_input_df <- df[, c("data")][, "user_input"] %>% as.data.table()
    dt <- cbind(dt, data_df_without_user_input)
    if (nrow(user_input_df) == 0) {
        return(dt)
    }
    cbind(dt, user_input_df)
}
```

**Comment by shankari** [comment]
in terms of scalability, we are going to allow users to specify a range. that is true irrespective of anything.
once the user specifies the range, we can be in one of several situations:
- **all trips and all locations from the range are within our limit (easy):** we can just load and display everything
- **even just the trip documents from the range are above the limit (easy):** we can just fail and tell people to reduce the range
- **trips is a high % of max documents, so only few locations could be loaded:** from experimenting with the label screen (https://github.com/e-mission/e-mission-docs/issues/226#issuecomment-762390537), we need a min of 100 points per trip for it to be recognizable and 500 would be better. This should be configurable but those are some ballpark estimates. So if trips + 100 * trips > max document count, we are in this situation.
   - We can load only the trips, blank out the map and insert a warning into the maps tab indicating that the range is too large to display trajectories.
   - We can also always have a blank map, only load trips, and have a button on the map tab to load the map. Then, we can warn users at that time that the range is too large to load trajectories. I think this is a bit easier from a UX perspective.
   - Instead of a blank map, when we are not loading trajectories, we could just display the starts and ends of the trips with a straight line between them.
- **trips + 100*trips < max documents and trips + locations > max documents:** downsample the trajectories. so we can just `$sample` with `N = max_documents - trips`

**Comment by shankari** [comment]
so if we use the option in which we always load only trips first, and then load trajectories from the map screen and a button, we would split this:
- **trips < max_documents:** load them
- **trips > max_documents:** warn user

when the user presses the "get trajectories" button:
- **trips + trajectories < max_documents:** get all locations
- else **trips + 100 * trips < max_documents:** get sample
- else **trips + 100 * trips > max_documents:** warn user that we can't do this

**Comment by asiripanich** [comment]
Closed with #49.
